EventFilter/CTPPSRawToDigi/plugins/CTPPSPixelDigiToRaw.cc

```diff
@@ -0,0 +1,200 @@
+#include "EventFilter/CTPPSRawToDigi/interface/CTPPSPixelDataFormatter.h"
+
+#include "CondFormats/DataRecord/interface/CTPPSPixelDAQMappingRcd.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/CTPPSPixelDAQMapping.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/CTPPSPixelFramePosition.h"
+
+#include "FWCore/PluginManager/interface/ModuleDef.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+//
+// class declaration
+//
+
+class CTPPSPixelDigiToRaw : public edm::stream::EDProducer<> {
+  public:
+    explicit CTPPSPixelDigiToRaw(const edm::ParameterSet&);
+    ~CTPPSPixelDigiToRaw() override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void beginStream(edm::StreamID) override;
+    void produce(edm::Event&, const edm::EventSetup&) override;
+    void endStream() override;
+
+    // ----------member data ---------------------------
+    unsigned long eventCounter_;
+    int allDigiCounter_;
+    int allWordCounter_;
+    bool debug_;
+    std::set<unsigned int> fedIds_;
+    std::string mappingLabel_;
+    edm::ESWatcher<CTPPSPixelDAQMappingRcd> recordWatcher_;
+    edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelDigi>> tCTPPSPixelDigi_;
+    std::vector<CTPPSPixelDataFormatter::PPSPixelIndex> v_iDdet2fed_;
+    CTPPSPixelFramePosition fPos_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+CTPPSPixelDigiToRaw::CTPPSPixelDigiToRaw(const edm::ParameterSet& iConfig) 
+{
+  //register your products
+  tCTPPSPixelDigi_ = consumes<edm::DetSetVector<CTPPSPixelDigi> >(iConfig.getParameter<edm::InputTag>("InputLabel"));
+
+  // Define EDProduct type
+  produces<FEDRawDataCollection>();
+  mappingLabel_ = iConfig.getParameter<std::string> ("mappingLabel");
+  // start the counters
+  eventCounter_ = 0;
+  allDigiCounter_ = 0;
+  allWordCounter_ = 0; 
+}
+
+
+CTPPSPixelDigiToRaw::~CTPPSPixelDigiToRaw()
+{
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+  void
+CTPPSPixelDigiToRaw::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  using namespace edm;
+  using namespace std;
+
+  eventCounter_++;
+
+  edm::Handle< edm::DetSetVector<CTPPSPixelDigi> > digiCollection;
+  iEvent.getByToken( tCTPPSPixelDigi_, digiCollection);
+
+  CTPPSPixelDataFormatter::RawData rawdata;
+  CTPPSPixelDataFormatter::Digis digis;
+
+  int digiCounter = 0;
+  for (auto const &di : *digiCollection) {
+    digiCounter += (di.data).size();
+    digis[ di.id] = di.data;
+  }
+  allDigiCounter_ += digiCounter;
+  edm::ESHandle<CTPPSPixelDAQMapping> mapping;
+  if (recordWatcher_.check( iSetup )) {
+    iSetup.get<CTPPSPixelDAQMappingRcd>().get(mapping);
+    for (const auto &p : mapping->ROCMapping)    {
+      CTPPSPixelDataFormatter::PPSPixelIndex iDdet2fed = { p.second.iD, p.second.roc, p.first.getROC(), p.first.getFEDId(), p.first.getChannelIdx() }; 
```

Comment:
if you define a constructor for `PPSPixelIndex`, you can just use `v_iDdet2fed_.emplace_back(p.second.iD, p.second.roc, p.first.getROC(), p.first.getFEDId(), p.first.getChannelIdx());` and avoid the temporary

Reply:
Ok